// postgraphile/build-turbo/postgraphile/pluginHook.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pluginHookFromOptions = exports.makePluginHook = void 0;
// @ts-ignore Allow importing JSON
const package_json_1 = require("../../package.json");
const graphql = require("graphql");
const identityHook = (input) => input;
const identityPluginHook = (_hookName, input, _options) => input;
function contextIsSame(context1, context2) {
    // Shortcut if obvious
    if (context1 === context2) {
        return true;
    }
    // Blacklist approach from now on
    const keys1 = Object.keys(context1);
    const keys2 = Object.keys(context2);
    if (keys1.length !== keys2.length) {
        return false;
    }
    // tslint:disable-next-line one-variable-per-declaration
    for (let i = 0, l = keys1.length; i < l; i++) {
        const key = keys1[i];
        if (context1[key] !== context2[key]) {
            return false;
        }
        if (!keys2.includes(key)) {
            return false;
        }
    }
    return true;
}
// Caches the last value of the hook, in case it's called with exactly the same
// arguments again.
function memoizeHook(hook) {
    let lastCall = null;
    return (argument, context) => {
        if (lastCall && lastCall.argument === argument && contextIsSame(lastCall.context, context)) {
            return lastCall.result;
        }
        else {
            const result = hook(argument, context);
            lastCall = {
                argument,
                context,
                result,
            };
            return result;
        }
    };
}
function shouldMemoizeHook(hookName) {
    return hookName === 'withPostGraphileContext';
}
function makeHook(plugins, hookName) {
    const combinedHook = plugins.reduce((previousHook, plugin) => {
        if (typeof plugin[hookName] === 'function') {
            return (argument, context) => {
                return plugin[hookName](previousHook(argument, context), context);
            };
        }
        else {
            return previousHook;
        }
    }, identityHook);
    if (combinedHook === identityHook) {
        return identityHook;
    }
    else if (shouldMemoizeHook(hookName)) {
        return memoizeHook(combinedHook);
    }
    else {
        return combinedHook;
    }
}
function makePluginHook(plugins) {
    const hooks = {};
    const emptyObject = {}; // caching this makes memoization faster when no context is needed
    function rawPluginHook(hookName, argument, context = emptyObject) {
        if (!hooks[hookName]) {
            hooks[hookName] = makeHook(plugins, hookName);
        }
        return hooks[hookName](argument, context);
    }
    const pluginHook = rawPluginHook('pluginHook', rawPluginHook, {});
    // Use this hook to check your hook is compatible with this version of
    // PostGraphile, also to get a reference to shared graphql instance.
    pluginHook('init', null, { version: package_json_1.version, graphql });
    return pluginHook;
}
exports.makePluginHook = makePluginHook;
function pluginHookFromOptions(options) {
    if (typeof options.pluginHook === 'function') {
        return options.pluginHook;
    }
    else {
        return identityPluginHook;
    }
}
exports.pluginHookFromOptions = pluginHookFromOptions;
